modules/core/src/matmul.dispatch.cpp

```diff
@@ -59,6 +60,138 @@ namespace cv
 *                                         GEMM                                           *
 \****************************************************************************************/
 
+#ifdef HAVE_CLBLAST
+
+// matD = alpha * matA * matB + beta * matC
+static bool ocl_gemm_clblast(InputArray matA, InputArray matB, double alpha,
+                             InputArray matC, double beta, OutputArray matD, int flags) {
+    int type = matA.type(), depth = CV_MAT_DEPTH(type), esz = CV_ELEM_SIZE(type);
+    bool haveC = matC.kind() != cv::_InputArray::NONE;
+    CV_CheckEQ(matB.type(), type, "Type of matB does not match the type of matA");
+    if (haveC) {
+        CV_CheckEQ(matC.type(), type, "Type of matC does not match the type of matA");
+    }
+    auto &device = ocl::Device::getDefault();
+    bool SupportFP64 = device.hasFP64(),
+         SupportFP16 = device.hasFP16();
+    if (!SupportFP64 && depth == CV_64F) {
+        return false;
+    }
+    if (!SupportFP16 && depth == CV_16F) {
+        return false;
+    }
+
+    Size sizeA = matA.size(), sizeB = matB.size(), sizeC = haveC ? matC.size() : Size(0, 0);
+    bool transA_ = (flags & GEMM_1_T) != 0, transB_ = (flags & GEMM_2_T) != 0, transC_ = (flags & GEMM_3_T) != 0;
+
+    if (transA_) {
+        sizeA = Size(sizeA.height, sizeA.width);
+    }
+    if (transB_) {
+        sizeB = Size(sizeB.height, sizeB.width);
+    }
+    if (haveC && transC_) {
+        sizeC = Size(sizeC.height, sizeC.width);
+    }
+
+    Size sizeD(sizeB.width, sizeA.height);
+
+    CV_CheckEQ(sizeA.width, sizeB.height, "Invalid dimension for matrix multiplification");
+    if (haveC) { // TODO: support matC broadcasting
+        CV_CheckTrue(sizeC == sizeD, "Shape of matC is not equal to the shape of matD");
+    }
+
+    matD.create(sizeD, type);
+    if (matA.offset() % esz != 0 || matA.step() % esz != 0 ||
+        matB.offset() % esz != 0 || matB.step() % esz != 0 ||
+        (haveC && (matC.offset() % esz != 0 || matC.step() % esz != 0)) )
+        return false;
```

Comment:
It makes sense to split this complex condition on dedicated `if` statements.

Reply:
Done